src/seraphis_mocks/mock_http_client_pool.cpp

```diff
@@ -0,0 +1,91 @@
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// NOT FOR PRODUCTION
+
+//local headers
+#include "mock_http_client_pool.h"
+
+//third party headers
+
+//standard headers
+
+//forward declarations
+
+
+namespace sp
+{
+namespace mocks
+{
+//-------------------------------------------------------------------------------------------------------------------
+epee::net_utils::http::abstract_http_client *ClientConnectionPool::acquire_unused_http_client(
+        const std::string &daemon_address,
+        const boost::optional<epee::net_utils::http::login> daemon_login,
+        const epee::net_utils::ssl_options_t ssl_support,
+        size_t &http_client_index)
+{
+    std::lock_guard<std::mutex> lock{m_http_client_pool_mutex};
+
+    // TODO: if m_http_client_pool.size() >= m_max_connections, wait until unused
+
+    bool found_unused_client = false;
+    for (size_t i = 0; i < m_http_client_pool.size(); ++i)
+    {
+        if (!m_http_client_pool[i].in_use)
+        {
+            http_client_index = i;
+            found_unused_client = true;
+            break;
+        }
+    }
+
+    if (!found_unused_client)
+    {
+        m_http_client_pool.emplace_back(pool_http_client_t{
+            .in_use = true,
+            .http_client = std::unique_ptr<epee::net_utils::http::abstract_http_client>(new net::http::client())
+        });
+        m_http_client_pool.back().http_client->set_server(daemon_address, daemon_login, ssl_support);
+        http_client_index = m_http_client_pool.size() - 1;
+    }
+
+    m_http_client_pool[http_client_index].in_use = true;
+    return m_http_client_pool[http_client_index].http_client.get();
+}
+//-------------------------------------------------------------------------------------------------------------------
+void ClientConnectionPool::release_http_client(size_t http_client_index)
+{
+    // Make the connection available for use again
```

Comment:
Will we ever prune the list of HTTP clients?

Reply:
I think it makes sense that once the scanner finishes a full pass, only 1 connection should be kept open, the rest can be closed. Seems good housekeeping.

In the future, I think this connection pool could potentially be a good class to build multi-daemon functionality with (1 client -> many different daemons), in which case it may make sense to keep connections to different daemons open. But that's probably a discussion for another time, just putting the thought out since I know you were interested in pushing that architecture forward